components/num-util/src/fixed_decimal.rs

```diff
@@ -0,0 +1,538 @@
+    /// even if the number has leading or trailing zeros.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use icu_num_util::FixedDecimal;
+    ///
+    /// let mut dec = FixedDecimal::from(120);
+    /// assert_eq!(0..=2, dec.magnitude_range());
+    /// ```
+    pub fn magnitude_range(&self) -> RangeInclusive<i16> {
+        self.lower_magnitude..=self.upper_magnitude
+    }
+
+    /// Shift the digits by a power of 10, modifying self.
+    ///
+    /// Leading or trailing zeros may be added to keep the digit at magnitude 0 (the last digit
+    /// before the decimal separator) visible.
+    ///
+    /// Can fail if the change in magnitude pushes the digits out of bounds; the magnitudes of all
+    /// digits should fit in an i16.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use icu_num_util::FixedDecimal;
+    ///
+    /// let mut dec = FixedDecimal::from(42);
+    /// assert_eq!("42", dec.to_string());
+    ///
+    /// dec.multiply_pow10(3).expect("Bounds are small");
+    /// assert_eq!("42000", dec.to_string());
+    /// ```
+    pub fn multiply_pow10(&mut self, delta: i16) -> Result<(), Error> {
+        match delta.cmp(&0) {
+            Ordering::Greater => {
+                self.upper_magnitude = self
+                    .upper_magnitude
+                    .checked_add(delta)
+                    .ok_or(Error::Limit)?;
+                // If we get here, then the magnitude change is in-bounds.
+                let lower_magnitude = self.lower_magnitude + delta;
+                self.lower_magnitude = cmp::min(0, lower_magnitude);
+            }
+            Ordering::Less => {
+                self.lower_magnitude = self
+                    .lower_magnitude
+                    .checked_add(delta)
+                    .ok_or(Error::Limit)?;
+                // If we get here, then the magnitude change is in-bounds.
+                let upper_magnitude = self.upper_magnitude + delta;
+                self.upper_magnitude = cmp::max(0, upper_magnitude);
+            }
+            Ordering::Equal => {}
+        };
+        self.magnitude += delta;
+        #[cfg(debug_assertions)]
+        self.check_invariants();
+        Ok(())
+    }
+
+    /// Shift the digits by a power of 10, consuming self and returning a new object if successful.
+    ///
+    /// Leading or trailing zeros may be added to keep the digit at magnitude 0 (the last digit
+    /// before the decimal separator) visible.
+    ///
+    /// Can fail if the change in magnitude pushes the digits out of bounds; the magnitudes of all
+    /// digits should fit in an i16.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use icu_num_util::FixedDecimal;
+    ///
+    /// let dec = FixedDecimal::from(42).multiplied_pow10(3).expect("Bounds are small");
+    /// assert_eq!("42000", dec.to_string());
+    /// ```
+    pub fn multiplied_pow10(mut self, delta: i16) -> Result<Self, Error> {
```

Comment:
hmm, one of the main values of having this variant of the method is that you could use it on an immutable `FixedDecimal` to produce a new one.
The way you use it here requires a mutable reference which diminishes the value of building a new one.

There are probably smart optimizations we can add later, but for now, maybe just make it immutable and `self.clone().multiply_pow10(delta)` ?

Reply:
The argument is `mut self`, not `&mut self`.  I think it is equivalent to an argument with type `self` with a line in the function such as `mut retval = self;`.  Is it confusing/bad style for an argument to be `mut self` but be by value and not reference?

Reply:
Ah, I misread the code. It's all good if you need it mutable inside.